### ISSUE TITLE
core: Can't use user variable in `keep_input_artifact`

### ISSUE THREAD

**Issue by styro** [hashicorp/packer#2630]
I've got the same issue as #1913 recurring in 0.8.5 when using a string user variable for the boolean keep_input_artifact in the vagrant post-processor.

Oddly enough though, I can use the same string syntax for booleans successfully in the headless setting in the virtualbox builder.

**Comment by cbednarski** [comment]
Thanks for the report. I've seen some options that want a boolean and others that want a string "true" or "false". I agree this is confusing and we should probably accept either one in any of these cases.

**Comment by styro** [comment]
I personally don't mind whether it's a string or a boolean, as long as at least one of them works :)

**Comment by dragon788** [comment]
Having this issue with 0.8.6 as well. Was just trying to be 'clever' and handle keeping the artifact based on whether I was in "dev" mode or "ci" mode and it appears I currently cannot do this due to the strict typing on keep_input_artifact.

**Comment by WyseNynja** [comment]
I'm running into this with 0.9.0 so I'm stuck on 0.7.5

**Comment by ferrarimarco** [comment]
I tried with packer 0.12.0 and this issue is still valid. Here is the log:
```
$ packer build ubuntu.json
Failed to parse template: 2 error(s) occurred:

* variable keep_input_artifact_post_processor: '' expected type 'string', got unconvertible type 'bool'
* post-processor 1.1: 1 error(s) decoding:

* 'keep_input_artifact' expected type 'bool', got unconvertible type 'string'
```

Here are the relevant parts of the template:
```
  "variables": {
    "keep_input_artifact": true,
  }
```

```
  "post-processors": [
    {
      "keep_input_artifact": "{{ user `keep_input_artifact` }}",
      "output": "ubuntu.box",
      "type": "vagrant",
    }
```
I am using boxcutter/ubuntu 16.04 template.

**Comment by rickard-von-essen** [comment]
@ferrarimarco Gist full log (`PACKER_LOG=1 packer build template.json`)

**Comment by ferrarimarco** [comment]
Here it is:
```
$ PACKER_LOG=1 packer build -only=virtualbox-iso -var-file=ubuntu1604.json ubuntu.json
2016/11/28 12:32:31 [INFO] Packer version: 0.12.0
2016/11/28 12:32:31 Packer Target OS/Arch: windows amd64
2016/11/28 12:32:31 Built with Go Version: go1.7.3
2016/11/28 12:32:31 Using internal plugin for qemu
2016/11/28 12:32:31 Using internal plugin for vmware-iso
2016/11/28 12:32:31 Using internal plugin for cloudstack
2016/11/28 12:32:31 Using internal plugin for googlecompute
2016/11/28 12:32:31 Using internal plugin for openstack
2016/11/28 12:32:31 Using internal plugin for file
2016/11/28 12:32:31 Using internal plugin for null
2016/11/28 12:32:31 Using internal plugin for amazon-ebs
2016/11/28 12:32:31 Using internal plugin for amazon-ebsvolume
2016/11/28 12:32:31 Using internal plugin for azure-arm
2016/11/28 12:32:31 Using internal plugin for parallels-pvm
2016/11/28 12:32:31 Using internal plugin for profitbricks
2016/11/28 12:32:31 Using internal plugin for virtualbox-iso
2016/11/28 12:32:31 Using internal plugin for virtualbox-ovf
2016/11/28 12:32:31 Using internal plugin for amazon-instance
2016/11/28 12:32:31 Using internal plugin for digitalocean
2016/11/28 12:32:31 Using internal plugin for parallels-iso
Failed to parse template: 1 error(s) occurred:
2016/11/28 12:32:31 Using internal plugin for amazon-chroot
2016/11/28 12:32:31 Using internal plugin for docker
2016/11/28 12:32:31 Using internal plugin for vmware-vmx

2016/11/28 12:32:31 Using internal plugin for windows-restart
2016/11/28 12:32:31 Using internal plugin for ansible
2016/11/28 12:32:31 Using internal plugin for chef-client
* post-processor 1.1: 1 error(s) decoding:
2016/11/28 12:32:31 Using internal plugin for chef-solo
2016/11/28 12:32:31 Using internal plugin for file
2016/11/28 12:32:31 Using internal plugin for puppet-server

2016/11/28 12:32:31 Using internal plugin for shell-local
2016/11/28 12:32:31 Using internal plugin for windows-shell
2016/11/28 12:32:31 Using internal plugin for ansible-local
2016/11/28 12:32:31 Using internal plugin for powershell
* 'keep_input_artifact' expected type 'bool', got unconvertible type 'string'
2016/11/28 12:32:31 Using internal plugin for puppet-masterless
2016/11/28 12:32:31 Using internal plugin for salt-masterless
2016/11/28 12:32:31 Using internal plugin for shell
2016/11/28 12:32:31 Using internal plugin for artifice
2016/11/28 12:32:31 Using internal plugin for docker-tag
2016/11/28 12:32:31 Using internal plugin for manifest
2016/11/28 12:32:31 Using internal plugin for amazon-import
2016/11/28 12:32:31 Using internal plugin for docker-import
2016/11/28 12:32:31 Using internal plugin for googlecompute-export
2016/11/28 12:32:31 Using internal plugin for vagrant-cloud
2016/11/28 12:32:31 Using internal plugin for atlas
2016/11/28 12:32:31 Using internal plugin for checksum
2016/11/28 12:32:31 Using internal plugin for compress
2016/11/28 12:32:31 Using internal plugin for docker-push
2016/11/28 12:32:31 Using internal plugin for shell-local
2016/11/28 12:32:31 Using internal plugin for vagrant
2016/11/28 12:32:31 Using internal plugin for docker-save
2016/11/28 12:32:31 Using internal plugin for vsphere
2016/11/28 12:32:31 Attempting to open config file: C:\Users\dev-ferrari\AppData\Roaming\packer.config
2016/11/28 12:32:31 [WARN] Config file doesn't exist: C:\Users\dev-ferrari\AppData\Roaming\packer.config
2016/11/28 12:32:31 Packer config: &{DisableCheckpoint:false DisableCheckpointSignature:false PluginMinPort:10000 PluginMaxPort:25000 Builders:map[amazon-chroot:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-builder-amazon-chroot vmware-vmx:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-builder-vmware-vmx null:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-builder-null amazon-ebsvolume:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-builder-amazon-ebsvolume virtualbox-iso:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-builder-virtualbox-iso virtualbox-ovf:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-builder-virtualbox-ovf amazon-instance:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-builder-amazon-instance cloudstack:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-builder-cloudstack parallels-pvm:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-builder-parallels-pvm profitbricks:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-builder-profitbricks docker:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-builder-docker qemu:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-builder-qemu openstack:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-builder-openstack file:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-builder-file amazon-ebs:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-builder-amazon-ebs digitalocean:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-builder-digitalocean vmware-iso:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-builder-vmware-iso googlecompute:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-builder-googlecompute azure-arm:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-builder-azure-arm parallels-iso:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-builder-parallels-iso] PostProcessors:map[shell-local:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-post-processor-shell-local vagrant:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-post-processor-vagrant artifice:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-post-processor-artifice amazon-import:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-post-processor-amazon-import checksum:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-post-processor-checksum docker-push:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-post-processor-docker-push docker-import:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-post-processor-docker-import vagrant-cloud:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-post-processor-vagrant-cloud atlas:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-post-processor-atlas compress:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-post-processor-compress docker-save:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-post-processor-docker-save vsphere:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-post-processor-vsphere docker-tag:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-post-processor-docker-tag manifest:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-post-processor-manifest googlecompute-export:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-post-processor-googlecompute-export] Provisioners:map[file:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-provisioner-file shell-local:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-provisioner-shell-local windows-restart:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-provisioner-windows-restart chef-client:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-provisioner-chef-client chef-solo:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-provisioner-chef-solo ansible-local:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-provisioner-ansible-local powershell:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-provisioner-powershell puppet-masterless:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-provisioner-puppet-masterless salt-masterless:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-provisioner-salt-masterless shell:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-provisioner-shell ansible:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-provisioner-ansible puppet-server:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-provisioner-puppet-server windows-shell:C:\Program Files\packer\packer.exe-PACKERSPACE-plugin-PACKERSPACE-packer-provisioner-windows-shell]}
2016/11/28 12:32:31 Setting cache directory: C:\Users\ferrarim\Workspaces\ubuntu\packer_cache
2016/11/28 12:32:31 ui error: Failed to parse template: 1 error(s) occurred:

* post-processor 1.1: 1 error(s) decoding:

* 'keep_input_artifact' expected type 'bool', got unconvertible type 'string'
2016/11/28 12:32:31 waiting for all plugin processes to complete...
```

**Comment by ferrarimarco** [comment]
@rickard-von-essen  I tried with the latest devevelopment version, no changes:
```
/opt # packer version
Packer v0.12.1.dev (e45dd0b3949f16ddc1447f25030b3fc69cd7a137)

/opt # packer validate ubuntu.json
Failed to parse template: 1 error(s) occurred:

* post-processor 1.1: 1 error(s) decoding:

* 'keep_input_artifact' expected type 'bool', got unconvertible type 'string'
```

**Comment by rickard-von-essen** [comment]
```
$ packer version
Packer v0.12.1.dev (17eda06d613aa18ae1da05c80800d5070b19ee11+CHANGES)

$ packer validate ubuntu.json
Template validated successfully.
```

Could you try on another system or in a vagrant box etc?

**Comment by ferrarimarco** [comment]
That log was generated from inside a Docker container, starting from [hashicorp/docker-hub-images Dockerfile-full](https://github.com/hashicorp/docker-hub-images/blob/master/packer/Dockerfile-full).

I just removed the `ENTRYPOINT` instruction because otherwise the container could not start at all.

**Comment by rickard-von-essen** [comment]
Ok, do you access the ubuntu repo through a volume mount into docker?

**Comment by ferrarimarco** [comment]
I tried in two ways:

1. Volume mount
1. Clone the repository directly in the container with Git, to exclude any mount side effects

**Comment by rickard-von-essen** [comment]
So what is the exact output of first:
```bash
docker run -it ubuntu:trusty bash
```

and then:
```bash
apt-get update > /dev/null && apt-get -y install git unzip wget > /dev/null

wget https://releases.hashicorp.com/packer/0.12.0/packer_0.12.0_linux_amd64.zip
unzip packer_0.12.0_linux_amd64.zip

git clone https://github.com/boxcutter/ubuntu.git && cd ubuntu && ../packer validate ubuntu.json
```

**Comment by ferrarimarco** [comment]
Here are the results with the "vanilla" template.

```
root@8ca1443a6c62:/# apt-get update > /dev/null && apt-get -y install git unzip wget > /dev/null
root@8ca1443a6c62:/# wget https://releases.hashicorp.com/packer/0.12.0/packer_0.12.0_linux_amd64.zip
--2016-11-30 07:58:17--  https://releases.hashicorp.com/packer/0.12.0/packer_0.12.0_linux_amd64.zip
Resolving releases.hashicorp.com (releases.hashicorp.com)... 151.101.120.69
Connecting to releases.hashicorp.com (releases.hashicorp.com)|151.101.120.69|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 10047746 (9.6M) [application/zip]
Saving to: 'packer_0.12.0_linux_amd64.zip'

100%[====================================================================================================================================================>] 10,047,746   777KB/s   in 9.5s   

2016-11-30 07:58:29 (1.01 MB/s) - 'packer_0.12.0_linux_amd64.zip' saved [10047746/10047746]

root@8ca1443a6c62:/# unzip packer_0.12.0_linux_amd64.zip  
Archive:  packer_0.12.0_linux_amd64.zip
  inflating: packer                  
root@8ca1443a6c62:/# git clone https://github.com/boxcutter/ubuntu.git && cd ubuntu && ../packer validate ubuntu.json
Cloning into 'ubuntu'...
remote: Counting objects: 2065, done.
remote: Compressing objects: 100% (44/44), done.
remote: Total 2065 (delta 18), reused 0 (delta 0), pack-reused 2019
Receiving objects: 100% (2065/2065), 372.84 KiB | 0 bytes/s, done.
Resolving deltas: 100% (1345/1345), done.
Checking connectivity... done.
Template validated successfully.
```

My edits to `ubuntu.json`:
```
  "variables": {
    "keep_input_artifact": "true",
  }
...
  "post-processors": [
    {
      "keep_input_artifact": "{{ user `keep_input_artifact` }}",
      ...
    }
```

Results:
```
root@8ca1443a6c62:/ubuntu# ../packer validate ubuntu.json
Failed to parse template: 1 error(s) occurred:

* post-processor 1.1: 1 error(s) decoding:

* 'keep_input_artifact' expected type 'bool', got unconvertible type 'string'
root@8ca1443a6c62:/ubuntu# 
```
Thanks @rickard-von-essen

**Comment by rickard-von-essen** [comment]
Check! I got a bit lost while trying to understand this.

I think a hint on why this doesn't work is that `keep_input_artifact` is the only bool which doesn't come from the plugins. It's defined in `template/template.go`.

**Comment by mwhooker** [comment]
hmm, was able to solve the validate error with this branch https://github.com/mitchellh/packer/tree/nokeepintpl

but it's untested, and I'm not sure the solution is valid. I'm not 100% sure where the `KeepInputArtifact` value is actually checked. Is it done globally or per-post-processor?

**Comment by ferrarimarco** [comment]
@mwhooker @rickard-von-essen is there something I should do/test? Thanks

**Comment by ferrarimarco** [comment]
@mwhooker Is there any update on this? How can I help? Thanks!

**Comment by mwhooker** [comment]
I'm afraid I'm not able to look into this right now. I investigated briefly last week, but didn't see any obvious solution. Will gladly look at a PR that solves this issue

**Comment by ferrarimarco** [comment]
@mwhooker what is the rationale behind [your commit](https://github.com/mitchellh/packer/commit/a11680717dee67d7bc8fe30df105cb5ed085a893)?

**Comment by ferrarimarco** [comment]
Tried with the latest packer (0.12.1), issue still here, thanks. @mwhooker

**Comment by dragon788** [comment]
@marco what if you remove the quotes surrounding the user variable macro
(the "{{" ) and leave them around the "true" where the variable is set?

**Comment by ferrarimarco** [comment]
@dragon788 If I understood correctly I should try with `"keep_input_artifact": "{{ user keep_input_artifact }}"`

Result was the same:
```
$packer validate ubuntu.json
Failed to parse template: 1 error(s) occurred:

* post-processor 1.1: 1 error(s) decoding:

* 'keep_input_artifact' expected type 'bool', got unconvertible type 'string'
```

**Comment by dragon788** [comment]
Hi Marco, actually I was thinking something more like the following (was replying from my phone earlier so was too tough to try and write out the code).

```
  "variables": {
    "keep_input_artifact": "true",
  }
...
  "post-processors": [
    {
      "keep_input_artifact": {{ user `keep_input_artifact` }},
      ...
    }
```

Packer (Go) is strongly typed, and that tends to bite users (myself included) when it comes to user variables.

Packer is checking user variables and apparently expects ALL of them to be text values, and then in this case you actually want your "keep_input_artifact" value in the template to be a boolean, so leaving out the quotes should leave just `"keep_input_artifact": true`, which is a Boolean rather than "true" which is a string.

**Comment by rickard-von-essen** [comment]
@dragon788, unfortunately that isn't even valid json.

**Comment by rickard-von-essen** [comment]
This is my analysis without looking much into the code.

1) First the template json decoded
2) Then each section (builders, post-processors, etc) split up per type.
3) User variables are substituted, and each plugin validates its part of the config.

The only boolean value that exists outside plugins are the generic `keep_input_artifact` which is handled as part of step 2). The fix in #1913 only applies to variables in 3).

**Comment by ferrarimarco** [comment]
Hi @rickard-von-essen and @mwhooker . Is there any update on a fix on this issue?

**Comment by rickard-von-essen** [comment]
Not more than that this is problem in the core code of packer which only impacts `keep_input_artifact` and currently it's not a high prio for us.

**Comment by ferrarimarco** [comment]
Hi. Can a fix for this be scheduled for 1.1?

Thanks!

**Comment by rickard-von-essen** [comment]
If there was a fix to schedule.

**Comment by jen20** [comment]
As an extra data point here @rickard-von-essen and @mwhooker: this does not only affect `keep_input_artifact` - it affects any boolean field which you might want to extract into a variable.

For example, take the following example (stripped down template):

```
{
    "variables": {
        "headless": false
    },
    //...... (if only JSON actually supported comments....)
    {
        "type": "vmware-iso",
        // More fields
        "headless": "{{ user `headless` }}"
    }
}
```

Running `packer validate template.json` here yields:

```
$ packer version
Packer v1.0.0

$ packer validate template.json
Failed to parse template: 1 error(s) occurred:

* variable headless: '' expected type 'string', got unconvertible type 'bool'
```

**Comment by mwhooker** [comment]
Happy to review PRs for this

**Comment by mwhooker** [comment]
To give some more context, yes this is a bug and I'm sorry we don't have a fix for this, but I don't have time to dig in any further than I already have. I've spent a bit of time on this and the fix isn't obvious to me. To me, the effort to fix this seems far greater than any possible work around, such as writing a preprocessor to write the variable out correctly. That, coupled with the fact that we intend to overhaul our configuration scheme in the near future bumps this down in the priority list for me.

If anyone wants to do the leg work to investigate what's going on and brainstorm possible solutions, I'd be happy to chat. In the meantime, though, this isn't a high priority for me.

**Comment by rickard-von-essen** [comment]
@jen20 your error is different, it complains that the user var value is not a string. If you surround your `false` with quotes `"false"` it works as expected. Example [Bento - Ubuntu 16.04 template](https://github.com/chef/bento/blob/master/ubuntu-16.04-amd64.json#L94).

I read the code quite carefully and this only effects booleans that are part of the templates _outside plugins_, and that is only `keep_input_artifact`. See https://github.com/hashicorp/packer/issues/2630#issuecomment-267674335

**Comment by shibumi** [comment]
I can confirm this bug with version 1.0.0 on Arch Linux

**Comment by curiositycasualty** [comment]
Still kickin' around in 1.0.3

**Comment by yves-vogl** [comment]
Still kickin' around in 1.3.1

**Comment by frankehlers** [comment]
Still kickin' around in 1.3.1

**Comment by SwampDragons** [comment]
Hopefully we'll be able to fix this when we implement the HCL2 configs; I'm not going to prioritize it before then because it's got a workaround. To paraphrase Matthew's comment above, It's a deep architectural quirk and the refactor necessary to make this one key work is dramatically more effort than implementing a workaround on your end.  For now I'm going to mark this with the "config-2.0" and "wontfix" flags to hopefully remove confusion about the maintainers' stance on this bug.

tl;dr We know it's still kickin' around and we aren't going to fix it independently of the large refactor we already have planned for late 2019.

**Comment by mschuchard** [comment]
I thought the HCL2 configs were already in Packer 1.3.x, since they were a prereq for the Vault integrations that @SwampDragons coincidentally implemented.

Additionally, will the HCL2 configs enable Arrays as well instead of the e.g. `"us-east-1,us-east-2"` workaround?

**Comment by SwampDragons** [comment]
HCL2 turned out not to be necessary for that Vault integration.

**Comment by SwampDragons** [comment]
I think whether we support arrays for particular options will have to do with the option in question.

**Comment by azr** [comment]
Hello there, just to say that this is now possible to set `keep_input_artifact` using variables in HCL2 ! I'm going to go ahead and close that issue ! Thanks !

**Comment by shibumi** [comment]
@azr can you give an example config file for visualizing how to do this?

**Comment by azr** [comment]
Of course, one thing we can do is use an input variable:

```hcl
variable "keep_that_input_artifact" {
 default = true
}

build {
  source = ["..."] # your sources
  
  # provision ...
  
  post-process "shell-local" {
    keep_input_artifact = var.keep_that_input_artifact
  }
}
```

That variable could be set from [cli arguments; env or a config file; ](https://www.packer.io/docs/from-1.5/variables#assigning-values-to-build-variables).

**Comment by azr** [comment]
Ah no, wait; @SwampDragons just found out that I was saying something wrong; the 'eval context' is not present at the moment we parse these variables and therefore we cannot use variables there het;


Issue is here: the second argument of decodebody should just take it as an argument:
https://github.com/hashicorp/packer/blob/cf514d31f2a0f80d8841307f0f2e6252f466fd22/hcl2template/types.build.post-processor.go#L25-L36

**Comment by divyamamgai** [comment]
I have the following post-processor.
```json
{
    "post-processors": [
        {
            "type": "compress",
            "output": "{{user `build_directory`}}/packer-XYZ-{{user `stack_name`}}-{{user `template_name`}}-{{build_type}}/XYZ-{{user `stack_name`}}-{{user `template_name`}}.tar.gz",
            "compression_level": "{{user `compression_level`}}",
            "keep_input_artifact": true
        }
    ],
}
```
I'm getting `<no-value>` in place of `{{build_type}}` (or `{{.BuilderType}}`) when `keep_input_artifact` is set to `true`. Is this behaviour documented somewhere and whats the workaround?

**Comment by nywilken** [comment]
Support for using variable interpolation in post-processors was add in the [1.7.3](https://github.com/hashicorp/packer/blob/master/CHANGELOG.md#173-june-15-2021) release via the change https://github.com/hashicorp/packer/pull/11094. I'm going to close this issue, as it can no longer be reproduced. If you are still running into issues please drop us a comment and we will gladly reopen. Cheers!
